dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/JdbcIconStore.java

```diff
@@ -0,0 +1,271 @@
+            """;
+
+    MapSqlParameterSource params = new MapSqlParameterSource();
+    params.addValue("description", icon.getDescription());
+    params.addValue(
+        "keywords", convertKeywordsSetToJsonString(icon.getKeywords().stream().toList()));
+    params.addValue("key", icon.getKey());
+
+    namedParameterJdbcTemplate.update(sql, params);
+  }
+
+  @Override
+  public List<Icon> getIcons(IconOperationParams params) {
+    String sql =
+        """
+              select c.iconkey as iconkey, c.description as icondescription, c.keywords as keywords, c.created as created, c.lastupdated as lastupdated,
+              c.fileresourceid as fileresourceid, c.createdby as createdby, c.custom as custom from icon c
+              """;
+
+    MapSqlParameterSource parameterSource = new MapSqlParameterSource();
+
+    sql = buildIconQuery(params, sql, parameterSource);
+
+    if (params.isPaging()) {
+      sql =
+          getPaginatedQuery(
+              params.getPager().getPage(), params.getPager().getPageSize(), sql, parameterSource);
+    }
+
+    return namedParameterJdbcTemplate
+        .queryForStream(sql, parameterSource, getIconRowMapper())
+        .toList();
+  }
+
+  private String buildIconQuery(
+      IconOperationParams params, String sql, MapSqlParameterSource parameterSource) {
+    SqlHelper hlp = new SqlHelper(true);
+
+    if (params.hasLastUpdatedStartDate()) {
+      sql += hlp.whereAnd() + " c.lastupdated >= :lastUpdatedStartDate ";
+
+      parameterSource.addValue(
+          ":lastUpdatedStartDate", params.getLastUpdatedStartDate(), Types.TIMESTAMP);
+    }
+
+    if (params.hasLastUpdatedEndDate()) {
+      sql += hlp.whereAnd() + " c.lastupdated <= :lastUpdatedEndDate ";
+
+      parameterSource.addValue(
+          "lastUpdatedEndDate", params.getLastUpdatedEndDate(), Types.TIMESTAMP);
+    }
+
+    if (params.hasCreatedStartDate()) {
+      sql += hlp.whereAnd() + " c.created >= :createdStartDate";
+
+      parameterSource.addValue("createdStartDate", params.getCreatedStartDate(), Types.TIMESTAMP);
+    }
+
+    if (params.hasCreatedEndDate()) {
+      sql += hlp.whereAnd() + " c.created <= :createdEndDate ";
+
+      parameterSource.addValue("createdEndDate", params.getCreatedEndDate(), Types.TIMESTAMP);
+    }
+
+    if (params.hasCustom()) {
+      sql += hlp.whereAnd() + " c.custom = :custom ";
+
+      parameterSource.addValue("custom", params.getCustom(), Types.BOOLEAN);
+    }
+
+    if (params.hasKeywords()) {
+
+      sql += hlp.whereAnd() + " keywords @> cast(:keywords as jsonb) ";
+
+      parameterSource.addValue("keywords", convertKeywordsSetToJsonString(params.getKeywords()));
+    }
+
+    if (params.hasKeys()) {
+      sql += hlp.whereAnd() + " c.iconkey IN (:keys )";
+
+      parameterSource.addValue("keys", params.getKeys());
+    }
+
+    if (params.hasSearch()) {
+      String searchValue = params.getSearch();
+
+      sql += hlp.whereAnd() + " c.iconkey ilike :search";
```

Comment:
Where did this requirement come from 🤔 ?

Reply:
may be @Birkbjo  can explain more. The client side needs `ilike` search for both keys and keywords. But for now we are only doing it for keys.

Reply:
So this comes from the fact that icons are now paginated.
Previously searching icons are done on the client. Because key and keywords filter only support full text matching, which is kind of useless for searching.

So for the client to be able to use this we kind of need a search. And this should've searched keywords as well, but that's a jsonb column and the function that trivially does that requires postgres13.